### ISSUE TITLE
Incorrect TCP sequence number ACK prematurely closing connection

### ISSUE THREAD

**Issue by JasonCoombs** [libbitcoin/libbitcoin-node#396]
This might be a problem with BOOST ASIO, or it could be an incorrectly-implemented strand. Perhaps we are failing to anticipate that ASIO result handlers may be called out of order?

The Wireshark capture suggests that the data for the next packet has not yet arrived on the wire at the time that libbitcoin-node generates an incorrect ACK for the next sequence number instead of an ACK for the current one as is expected and required.

<img width="2518" alt="screen shot 2018-11-10 at 12 54 26 pm" src="https://user-images.githubusercontent.com/16699430/48307029-5d14f400-e4e8-11e8-85d2-c86e84fd0039.png">

After sending another packet which Wireshark labels `[TCP Windows Update]` to attempt to resync sequence numbers with the peer, we immediately close the TCP connection `[FIN, ACK]` without waiting for a response.

This is observed with one thread in the network threadpool.

**Comment by JasonCoombs** [comment]
Here's another example of a bad decision to close a connection on a high-performing, responsive peer

<img width="2126" alt="screen shot 2018-11-12 at 6 40 15 pm" src="https://user-images.githubusercontent.com/16699430/48391415-9b5a1100-e6aa-11e8-9013-e9050ffb9464.png">

The console output says we've decided to terminate the connection because `protocol_ping_timer` followed IMMEDIATELY by `Ping latency limit exceeded` which is absurd, this is NOT a ping timeout.

```
04:32:38.301851 VERBOSE [network] Received block from [144.76.27.145:18333] (1954 bytes)
04:32:38.303953 DEBUG [database] Output cache hit rate: 0.399813, size: 9998
04:32:38.304390 INFO [node] Validated #93782 [000000000033e3e249965048cc5f860fd49d6c73424f3a80b085b8ea0653b202]
04:32:38.304987 VERBOSE [network] Received block from [144.76.27.145:18333] (868 bytes)
04:32:38.305114 VERBOSE [network] Fired protocol_ping timer on [144.76.27.145:18333] success
04:32:38.305169 DEBUG [network] Ping latency limit exceeded [144.76.27.145:18333]
04:32:38.306642 DEBUG [database] Output cache hit rate: 0.399816, size: 10000
04:32:38.306996 INFO [node] Validated #93783 [000000009877d45a5eda16d141555d54c58b3dd8383121b0672e2e716289f24d]
04:32:38.307389 VERBOSE [network] Stop protocol_version on [144.76.27.145:18333] connection timed out
04:32:38.307467 DEBUG [network] Outbound channel stopped [144.76.27.145:18333] success
04:32:38.307533 VERBOSE [network] Connecting to [176.233.66.60:18333]
```

**Comment by JasonCoombs** [comment]
I don't understand why you're being combative and non-cooperative. Do you not want any other developer to work on this project? My fork of libbitcoin is proceeding as planned now, so my efforts will just move to my repository instead. But it's counterproductive for you to refuse my good faith help.

**Comment by JasonCoombs** [comment]
So it looks like the callback function is being invoked immediately, or twice, and _pending is true:

https://github.com/libbitcoin/libbitcoin-network/blob/master/src/protocols/protocol_ping_60001.cpp#L61

This could be another instance where we have incorrectly presumed that ASIO callback functions are going to be called in a particular expected order sequence but the documentation says we must expect callbacks at any time with no particular ordering. Firing the callback immediately after dispatching a ping send doesn't even give time to send the ping before the timeout is detected.

**Comment by JasonCoombs** [comment]
Another case where a timer fires, callback is queued, and result handler invoked in 0.000065 seconds:

```
19:15:16.400204 VERBOSE [network] Fired protocol_header_in timer on [95.216.35.252:18333] success
19:15:16.400269 DEBUG [node] Peer [95.216.35.252:18333] is more behind or exceeded configured header latency.
```

the result is an incorrect peer disconnect during a perfectly-fine high-performing peer connection in which a current or correct next batch of headers is being received.

**Comment by evoskuil** [comment]
When any timer fires its callback is always invoked on the next availbke thread. Shorter delays for this are desirable.

Whether it is “incorrect” or not has nothing to do with the time to invoke the timer’s handler.

**Comment by evoskuil** [comment]
Strand: 
https://github.com/libbitcoin/libbitcoin/blob/master/include/bitcoin/bitcoin/utility/work.hpp#L142

**Comment by evoskuil** [comment]
> In fact there is never even a ping sent on the wire. Surely it's a bug to imagine we have sent a ping and to give up on the peer because we didn't get a reply to our imaginary ping after 0.000055 seconds.

You are again assuming bad behavior from an entirely superficial and incorrect understanding of actual behavior. The latency is *handled* in the ping timer handler. The ping time event is not when the ping message is sent but when the timer handler is evaluated.

**Comment by evoskuil** [comment]
> and _pending is already set to true:

Of course, which is established when the ping message is sent.

**Comment by JasonCoombs** [comment]
> Of course, which is established when the ping message is sent, which first happens immediately upon ping protocol start().

No. Read more carefully instead of just assuming that I'm wrong. What happens is that no ping is ever sent but the result handler is immediately invoked and because _pending is already set to true the decision is made to close the connection with the peer. There might be a ping queued for sending but it looks like that isn't happening in priority to the ping result handler being invoked too quickly. There is no reason at all for the connection to be closing when this race condition results in the result handler being invoked before the ping is sent.

**Comment by evoskuil** [comment]
The ping has been queued to send. Whether it is received at that point is in part up to the peer. In other words, waiting to consider it sent until the peer decided to accept it gives the peer control of the channel. Nor’easter your timeout.

**Comment by evoskuil** [comment]
You are mistaken about the behavior of the timer. It is the *expiration* of the time period. It fires st timeout.

**Comment by JasonCoombs** [comment]
no ping has been sent, so if the protocol_ping timer is fired after a timeout then please show me what keeps track of the time because I'm not seeing it, and also note that the `Ping latency limit exceeded` message comes not from the timer result handler but from the call to `protocol_ping_60001::send_ping(const code& ec)`

there must have been a prior call to `send_ping()` in order for `_pending` to be true, but it's the `handle_receive_pong` result handler that is supposed to set `_pending` to false. what makes you think that we are certain to get that result handler invoked before the queued and constantly-updating set of network I/O read events?

I asked before where do we look to see the queue of events, the set of result handlers and actions that are waiting for ASIO to dispatch on its threadpools. we need a better way to see what the queue contains, and if you will show me where that lives in memory or what source code manages that queue this will become easier to debug.

**Comment by JasonCoombs** [comment]
in summary, there does not appear to be any actual timer keeping track of the time since the ping was sent, there doesn't appear to be any confirmation being checked to see if the ping actually was sent, and I don't think the ping processing event queue is using the same strand as header I/O uses or if it is using the same strand then perhaps the ping send event hasn't been processed yet precisely because the strand is postponing it, and some combination of these factors results in no ping having been sent at the moment when a new call to `send_ping()` at which point `_pending` is still set to true and the timeout error is inferred when one has not in fact happened. finally, we shouldn't be pinging at all if the peer is doing a perfectly fine job of sending us headers, blocks, or other data.

**Comment by JasonCoombs** [comment]
I understand now that the `protocol_ping timer` is being invoked at the expiration of the timer, or because of a timeout condition, although I still don't see where the timer is actually timing anything relative to the `send_ping` event. after triple checking with three separate occurrences of this failure scenario (unwanted and incorrect termination of a high-performing peer connection which would have continued providing headers, addresses and blocks if we didn't disconnect) it is very clear that if the ping timer was not activated in error then there must be a `send_ping` event pending which has not resulted in any network activity yet.

most likely, the `send_ping` is using the same strand as the rest of the queued network events, or even sharing a single strand with all queued events of every type, and as I understand this from reading the ASIO documentation and various blogs and GitHub repositories where this question is discussed, we are supposed to have separate strands for each activity that we want to allow in parallel for which we need to enforce a particular sequence.

I believe it is incorrect to prevent the `send_ping` event from occurring until after all incoming headers and blocks have been received and processed, and then to trigger `protocol_ping timer` and determine thereby that the peer has timed out on its expected pong response to our ping, when we have not yet sent a ping for the peer to pong!

until there's a better way to inspect the content of the ASIO queue at runtime, I'm inserting additional logging to conclusively determine whether there is a `send_ping` event queued which has not yet been invoked via ASIO threadpool result handler / callback function, because I believe it's this queued event that our timer is telling us incorrectly has timed out indicating a low-performing or broken peer.

if this is correct, and we are going to continue to attach every event to the same strand, then we must not start these event timers until their corresponding network events have successfully transmitted data to the peer.

**Comment by JasonCoombs** [comment]
> You are just guessing about everything. I recommend you spend some time learning how it actually works before submitting bugs.

my analysis is not wrong. the bug exists, and my misunderstanding of the purpose of the ping timer is understandable when you consider that the packet captures prove conclusively that no ping has been sent to the peer at the point where we are deciding that there has been a timeout waiting for the peer to respond to the ping. there's no reason to wait to open Issues to report a bug when it is clear there is an issue that needs further debugging and an eventual fix. why don't you spend your time writing useful technical responses instead of reacting to every open Issue as though it threatens your livelihood.

**Comment by JasonCoombs** [comment]
Presumably this bug has still not been fixed? I would like to submit the fix, if you are interested.

**Comment by evoskuil** [comment]
Given the preceding discussion I’m not aware of any bug. But you are welcome to submit a PR.

**Comment by JasonCoombs** [comment]
I didn't want to argue with you. The problem is entirely in our code, not a result of peer behavior. I will submit a PR and you can see how my fix resolves the problem by adding the missing step to the timer callback queue mechanism.

**Comment by evoskuil** [comment]
I haven’t said that it isn’t, only that you have only demonstrated expected behavior above.

**Comment by JasonCoombs** [comment]
You will need to help decide what to do about the change of behavior that the bug fix causes. Right now the decision being made to terminate incorrectly because of the queue management bug results in libbitcoin searching for other peers more often than necessary. More testing will need to be done to determine if the bug fix introduces a new vulnerability to eclipse attacks. I agree that there is some benefit to switching peers frequently, but right now it's happening by mistake not by design.

https://www.usenix.org/node/190891

**Comment by evoskuil** [comment]
Channels have configurable timeouts, this is not an issue.

**Comment by JasonCoombs** [comment]
Wait to see the fix. You're having trouble understanding that no data has been sent to the peer yet, but the configurable timeout expires and the peer is deemed unworthy because it hasn't responded to the ping before the timeout. But we haven't even sent the ping to it yet. This is an issue.

**Comment by evoskuil** [comment]
We had this discussion, and as I pointed out this is insufficient to declare a problem. It would help if you could show specifically what is the issue as opposed to relying on message timing. Code may help.

**Comment by JasonCoombs** [comment]
PR here: https://github.com/libbitcoin/libbitcoin-network/pull/239
